lib/semantics/check-omp-structure.cc

```diff
@@ -146,15 +147,31 @@ void OmpStructureChecker::Leave(const parser::OmpClauseList &) {
     }
 
     if (auto *clause{FindClause(OmpClause::ORDERED)}) {
-      if (FindClause(OmpClause::LINEAR)) {
-        // only one ordered clause is allowed
-        const auto &orderedClause{
-            std::get<parser::OmpClause::Ordered>(clause->u)};
-        if (orderedClause.v.has_value()) {
+      // only one ordered clause is allowed
+      const auto &orderedClause{
+          std::get<parser::OmpClause::Ordered>(clause->u)};
+
+      if (orderedClause.v.has_value()) {
+        if (FindClause(OmpClause::LINEAR)) {
           context_.Say(clause->source,
               "A loop directive may not have both a LINEAR clause and "
               "an ORDERED clause with a parameter"_err_en_US);
         }
+
+        if (auto *clause2{FindClause(OmpClause::COLLAPSE)}) {
+          const auto &collapseClause{
+              std::get<parser::OmpClause::Collapse>(clause2->u)};
+          // ordered and collapse both have parameters
+          const auto orderedValue{GetIntValue(orderedClause.v)};
```

Comment:
Whenever you have

```
const auto foo{...};
if (foo) {
  ...
}
```

it can be made more readable by placing the declaration into the predicate expression.

```
if (const auto orderedValue{GetIntValue(orderedClause.v)}) {
  if (const auto collapseValue{GetIntValue(collapseClause.v)}) {
    if (*orderedValue > 0 && *orderedValue < *collapseValue) {
      Say("you lose");
    }
  }
}
```

Reply:
Thank you!